### PR TITLE
Fix bugs/warnings reported by ccc-analyzer

### PR DESCRIPTION
**Pull request by jjnicola**
**What**:
Fix memory leaks, and other stuff that ccc-analyzer doesn't like
Jira: SC-695

**Why**:

**How**:

**Checklist**:

- [ ] Tests
- [x] PR merge commit message adjusted

### CODE REVIEW

File: misc/ipc.c [hunk 1]
```diff
@@ -71,8 +71,7 @@ ipc_destroy (struct ipc_context *context)
       rc = ipc_pipe_destroy (context->context);
```
Comment (Kraemii) [inline misc/ipc.c:71]: Use an if statement instead of a switch with a single case

**Review by ArnoStiefvater — CHANGES_REQUESTED** [review]
Empty change request to prevent accidental merge. (Should be released for GOS 22.4.5 and not before that)

File: misc/ipc_pipe.c [hunk 2]
```diff
@@ -73,7 +73,8 @@ ipc_pipe_retrieve (struct ipc_pipe_context *context)
   fcntl (rfd, F_SETFL, pf | O_NONBLOCK);
   if ((result = calloc (1, IPC_MAX_BUFFER)) == NULL)
     return NULL;
-  if ((rr = read (rfd, result, IPC_MAX_BUFFER)) > 0)
+  rr = read (rfd, result, IPC_MAX_BUFFER);
+  if (rr > 0)
```
Comment (Kraemii) [inline misc/ipc_pipe.c:77]:
```suggestion
  if ( read (rfd, result, IPC_MAX_BUFFER) > 0)
```
The value of `rr` is only used here, so the return value of the function can be used directly and `rr` is not needed anymore

File: nasl/exec.c [hunk 3]
```diff
@@ -1403,7 +1406,7 @@ nasl_exec (lex_ctxt *lexic, tree_cell *st)
                   s3 = g_malloc0 (sz + 1);
                   if (p - p1 > 0)
                     memcpy (s3, p1, p - p1);
-                  if (sz > p - p1)
+                  if (sz > p - p1 && p != NULL)
```
Comment (Kraemii) [inline nasl/exec.c:1409]:
```suggestion
                  if (p != NULL && sz > p - p1)
```

File: nasl/nasl_host.c [hunk 4]
```diff
@@ -554,15 +554,15 @@ nasl_same_host (lex_ctxt *lexic)
 
   for (i = 0; i < 2; i++)
     g_free (a[i]);
-  if (cmp_hostname)
+  // if (cmp_hostname)
+  //  {
```
Comment (Kraemii) [inline nasl/nasl_host.c:558]:
```suggestion
```

File: nasl/nasl_host.c [hunk 5]
```diff
@@ -554,15 +554,15 @@ nasl_same_host (lex_ctxt *lexic)
...
-        {
-          for (j = 0; j < names_nb[i]; j++)
-            g_free (names[i][j]);
-          g_free (names[i]);
-        }
+      for (j = 0; j < names_nb[i]; j++)
+        g_free (names[i][j]);
+      g_free (names[i]);
     }
+  // }
```
Comment (Kraemii) [inline nasl/nasl_host.c:565]:
```suggestion
```

File: nasl/nasl_ssh.c [hunk 6]
```diff
@@ -1064,7 +1064,8 @@ nasl_ssh_login_interactive (lex_ctxt *lexic)
     {
       /* Our strategy for kbint is to send the password to the first
          prompt marked as non-echo.  */
-      while ((rc = ssh_userauth_kbdint (session, NULL, NULL)) == SSH_AUTH_INFO)
+      rc = ssh_userauth_kbdint (session, NULL, NULL);
+      while (rc == SSH_AUTH_INFO)
```
Comment (Kraemii) [inline nasl/nasl_ssh.c:1068]:
```suggestion
      while (ssh_userauth_kbdint (session, NULL, NULL) == SSH_AUTH_INFO)
```
We actually never use the value of `rc`, except for the while condition. I guess we can remove `rc` completely and put the function directly into the while loop

File: nasl/nasl_ssh.c [hunk 7]
```diff
@@ -1090,6 +1091,7 @@ nasl_ssh_login_interactive (lex_ctxt *lexic)
               if (s && *s && !echoflag && !found_prompt)
                 goto leave;
             }
+          rc = ssh_userauth_kbdint (session, NULL, NULL);
```
Comment (Kraemii) [inline nasl/nasl_ssh.c:1094]: We do not need this then either

File: src/openvas.c [hunk 8]
```diff
@@ -554,7 +554,8 @@ openvas (int argc, char *argv[], char *env[])
     }
   tzset ();
 
-  if ((err = init_logging ()) != 0)
+  err = init_logging ();
+  if (err != 0)
```
Comment (Kraemii) [inline src/openvas.c:558]:
```suggestion
  if (init_logging () != 0)
```
we do not need to save the err here, as we overwrite it anyways

File: src/openvas.c [hunk 9]
```diff
@@ -607,13 +608,16 @@ openvas (int argc, char *argv[], char *env[])
       globals = g_malloc0 (sizeof (struct scan_globals));
       globals->scan_id = g_strdup (get_scan_id ());
 
-      if ((err = attack_network_init (globals, config_file)) != 0)
-        return EXIT_FAILURE;
-
+      err = attack_network_init (globals, config_file);
+      if (err != 0)
```
Comment (Kraemii) [inline src/openvas.c:612]:
```suggestion
      if (attack_network_init (globals, config_file) != 0)
```

File: nasl/lint.c [hunk 10]
```diff
@@ -843,27 +847,29 @@ nasl_lint (lex_ctxt *lexic, tree_cell *st)
...
 
   /* Loads all defined functions. */
-  if ((ret = nasl_lint_def (lexic_aux, st, lint_mode, &include_files,
-                            &func_fnames_tab, err_fname, &called_funcs,
-                            &def_func_tree))
-      == NULL)
+  ret =
+    nasl_lint_def (lexic_aux, st, lint_mode, &include_files, &func_fnames_tab,
+                   err_fname, &called_funcs, &def_func_tree);
+  if (ret == NULL)
```
Comment (Kraemii) [inline nasl/lint.c:865]:
```suggestion
  if (nasl_lint_def (lexic_aux, st, lint_mode, &include_files, &func_fnames_tab,
                   err_fname, &called_funcs, &def_func_tree) == NULL)
```
The value of `ret` is overwritten anyways and is not used before that, so it is not necessary to save it here

File: nasl/lint.c [hunk 11]
```diff
@@ -880,10 +886,9 @@ nasl_lint (lex_ctxt *lexic, tree_cell *st)
 
   /* Now check that each function was loaded just once. */
   lint_mode = 0;
-  if ((ret =
-         nasl_lint_def (lexic, st, lint_mode, &include_files, &func_fnames_tab,
-                        err_fname, &called_funcs, &def_func_tree))
-      == NULL)
+  ret = nasl_lint_def (lexic, st, lint_mode, &include_files, &func_fnames_tab,
+                       err_fname, &called_funcs, &def_func_tree);
+  if (ret == NULL)
```
Comment (Kraemii) [inline nasl/lint.c:891]:
```suggestion
  if (nasl_lint_def (lexic, st, lint_mode, &include_files, &func_fnames_tab,
                       err_fname, &called_funcs, &def_func_tree) == NULL)
```
and here

**Comment by jjnicola** [comment]
Hey @Kraemii !!! thanks a lot for the review!

**Review by nichtsfrei — DISMISSED** [review]

File: nasl/lint.c [hunk 12]
```diff
@@ -843,26 +847,28 @@ nasl_lint (lex_ctxt *lexic, tree_cell *st)
...
-    /* FAKE_CELL if success, NULL otherwise which counts as error */
-    if ((ret = check_description_block (lexic_aux, desc_block)) == NULL)
-      {
-        inc_errors_cnt ();
-      }
-
+    {
+      /* FAKE_CELL if success, NULL otherwise which counts as error */
+      ret = check_description_block (lexic_aux, desc_block);
+      if (ret == NULL)
```
Comment (Kraemii) [inline nasl/lint.c:853]:
```suggestion
      if (check_description_block (lexic_aux, desc_block) == NULL)
```

File: nasl/charcnv.c [hunk 13]
```diff
@@ -91,7 +91,7 @@ check_dos_char_slowly_ntlmssp (uint16 c)
 
   len1 = convert_string_ntlmssp (CH_UTF16LE, CH_DOS, &c, 2, buf, sizeof (buf),
                                  False);
-  if (len1 == 0)
+  if (len1 <= 0)
```
Comment (nichtsfrei) [inline nasl/charcnv.c:94]: `convert_string_ntlmssp` is defined as `size_t` which is unsigned?

**Review by nichtsfrei — CHANGES_REQUESTED** [review]

File: nasl/charcnv.c (hunk 13 above)
Comment (jjnicola) [reply nasl/charcnv.c:94]: making len1 and len2 `size_t` var, ccc still complains.

Comment (nichtsfrei) [reply nasl/charcnv.c:94]: yes `convert_string_ntlmssp` is using `(size_t) -1` as an error code and casting to int will get -1 so I'm wrong and `<= 0` is correct.

File: nasl/charcnv.c [hunk 14]
```diff
@@ -87,11 +87,14 @@ check_dos_char_slowly_ntlmssp (uint16 c)
...
-  int len1, len2;
+  size_t len1, len2;
 
   len1 = convert_string_ntlmssp (CH_UTF16LE, CH_DOS, &c, 2, buf, sizeof (buf),
                                  False);
-  if (len1 == 0)
+
+  /* convert_string_ntlmssp returns a size_t value, and uses
+  * (size_t) -1 as error code */
+  if (len1 == 0 || len == (size_t) -1)
```
Comment (Kraemii) [inline nasl/charcnv.c:97]:
```suggestion
  if (len1 == 0 || len1 == (size_t) -1)
```

**Review by ArnoStiefvater — APPROVED** [review]

**Review by Kraemii — APPROVED** [review]

**Review by nichtsfrei — APPROVED** [review]

**Pull request merged**